server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceComputer.java

```diff
@@ -203,7 +204,8 @@ public DesiredBalance compute(
                         && routingAllocation.deciders()
                             .canAllocate(shardRouting, routingNode, routingAllocation)
                             .type() != Decision.Type.NO) {
-                        final var shardToInitialize = unassignedPrimaryIterator.initialize(nodeId, null, 0L, changes);
+                        final long expectedShardSize = getExpectedShardSize(shardRouting, 0L, routingAllocation);
+                        final var shardToInitialize = unassignedPrimaryIterator.initialize(nodeId, null, expectedShardSize, changes);
```

Comment:
Likewise, reverting this change doesn't make any `DesiredBalanceComputerTests` fail.